all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,526 @@
+environment variable can be set to point to it.
+```
+
+#### Possible CLI Creation Flow
+
+Assuming cloud support is present in `temporal` CLI, you can have the normal certificate generation steps, e.g.
+
+```
+temporal cloud gen ca ...
+temporal cloud gen leaf ...
+temporal cloud namespace add-accepted-client-ca ...
+```
+
+And that last step can by default set a `temporal-cloud` profile in the configuration with the address, namespace, and
+mTLS certificates. So now all you have to do to use cloud in CLI is either `--profile temporal-cloud` or set
+`TEMPORAL_PROFILE=temporal-cloud` environment variable. Granted you may want better certificate management than that and
+whether this is done by default can be discussed, but the idea is the same.
+
+#### Multi-Namespace Config Use
+
+Some users may want the same configuration for multiple namespaces (or multiple addresses or multiple API keys or
+whatever). Users can supply partial configuration. For instance, you can have this config:
+
+```toml
+[profile.temporal-cloud]
+address = "my-dev-ns.a1b2c.tmprl.cloud:7233"
+tls.client_cert_path = "path/to/my/dev-cert.pem"
+tls.client_key_path = "path/to/my/dev-cert.pem"
+```
+
+And then still use CLI like
+
+    temporal workflow list --namespace my-specific-namespace
+
+This will use the config file information for all but the namespace. Similarly you could just set the
+`TEMPORAL_NAMESPACE` env var to `my-specific-namespace` and run:
+
+    temporal workflow list
+
+Similarly, if you're using an SDK, you can load the config and just switch namespaces in code. So you can have:
+
+```python
+config = envconfig.load_client_connect_config()
+client = await Client.connect(namespace="my-specific-namespace", **config)
+```
+
+## Specification
+
+### Values and File Format
+
+* The format is a TOML object with the following fields:
+  * `profile` - Object with keys as profile names. The default profile should be named `default`.
+    * `<name>` - Profile object with configuration values. Profile names are case-insensitive. 💭 Why? They are needed
+      for environment variables which are case insensitive. It is a validation error to have a config file with two
+      separately-cased profile names that are equal case-insensitively.
+      * `address` - Client address, aka gRPC "host:port". This cannot be a URL, it must be `host:port`.
+      * `namespace` - Client namespace.
+      * `api_key` - Client API key.
+      * `tls` - A boolean (true is same as empty object) _or_ an object. Note the default for this value is dependent
+        upon other settings here. 💭 Why? We regret not making TLS the default in the past, so we will make it default
+        if `api_key` is present. If TLS is an object, it can have the following possible fields:
+        * `client_cert_path` - File path to mTLS client cert. Mutually exclusive with `client_cert_data`.
```

Comment:
For all path options:
 - If relative, evaluates relative to the config file?
 - Is there anything particular that we should specify about paths on Windows, to avoid cross-languages incoherencies? Should we be worried about WSL and similar?

Don't necessarily need to write this just here, but please put it somewhere so that it survives the review.

Reply:
> If relative, evaluates relative to the config file?

Good question. I generally like relative to mean current working directory, but I understand that's confusing for external configs. I will clarify that this should be relative to config file.

> Is there anything particular that we should specify about paths on Windows, to avoid cross-languages incoherencies? Should we be worried about WSL and similar?

Nope, not that I can think of

Reply:
There are three ways this path setting can be set: config file, env var, and programmatically. After some thought, it is confusing to have different ways of specifying the same value have different relative bases, so I think it makes the most sense to be all relative to the current working directory. I will add a point clarifying this.

Reply:
Treating these config keys different based on where they are set is confusing IMO. We have a want in our SDKs to be able to load the environment config as an object to be viewed. We want that object to have the _exact_ values the user set, not some resolved form (just in case they want to see what was there or want to save or whatever). How are you going to have different rules for the same config key from different places and still give a full config object with overrides?

Reply:
I am strongly in favor of relative paths in the config file being relative to the config file, and not the user's PWD. I think the extra implementation complexity of resolving the path differently depending on whether it comes from the config file or the environment is acceptable.

Putting a relative path in a config file, and then interpreting that path relative to the user's PWD at runtime, would be extremely surprising and confusing to most config file users IMO. This is not how most config-file-using apps I am familiar with behave.

It also makes redistribution of config files much harder, because you cannot include supplementary files alongside them without having some special logic somewhere to rewrite paths etc. You can no longer just zip up a config file and corresponding TLS keys etc. and send it to someone, you have to have the recipient rewrite the paths etc. to work reliably.

Reply:
As a compromise, I propose _requiring_ absolute paths and rejecting relative paths as invalid config. We can then revisit this point later.

I do agree that users should have programmatic access to the as-written values in the env and config file. However, I think my concern about relative paths is serious enough that it does need to be addressed one way or another.

Reply:
> I propose requiring absolute paths and rejecting relative paths as invalid config. We can then revisit this point later.

I assume you are not proposing this for `--tls-cert-path` that already exists on CLI today? We already support relative paths in `TEMPORAL_TLS_CERT`, are you proposing when we deprecate this in favor of `TEMPORAL_TLS_CLIENT_CERT_PATH` we remove the ability to use relative paths there? Also, are you proposing when we deprecate `temporal env --key tls-cert-path --value my/relative/path.cert` that it's replacement of `temporal config --key tls.client-cert-path --value my/relative/path.cert` won't work anymore?

Reply:
I am proposing that, in the config, we require absolute paths. I am not proposing any other changes (certainly no changes to the behavior of the existing env variables or flags).

Reply:
> I am proposing that, in the config, we require absolute paths

Ah, ok I misread "As a compromise, I propose requiring absolute paths and rejecting relative paths as invalid config" as all config, but you mean config _file_. I think we can mostly do this.

So, concerning the last part of my comment above, can you clarify what you expect the behavior of `temporal config --key tls.client-cert-path --value my/relative/path.cert` to be that writes to config file (compared to the deprecated form of `temporal env --key tls-cert-path --value my/relative/path.cert`)? Before, our env config supported relative paths, but now our new one does not, which I assume is an intentional regression?

Reply:
Yep, I think we would have to expand the path to its absolute form before writing to the config file, unless we want to simply fail saying the path is invalid/must be absolute.

Reply:
Ok, I will update this doc clarifying that config file _loading_ should fail if the path is not absolute (for now) and maybe config file writing from CLI can resolve the path.